src/eval/operation.rs

```diff
@@ -604,48 +614,56 @@ impl<R: ImportResolver, C: Cache> VirtualMachine<R, C> {
                     .pop_arg(&self.cache)
                     .ok_or_else(|| EvalError::NotEnoughArgs(2, String::from("generate"), pos_op))?;
 
-                if let Term::Num(n) = *t {
-                    let n_int = n as usize;
-                    if n < 0.0 || n.fract() != 0.0 {
-                        Err(EvalError::Other(
-                            format!(
-                            "generate: expected the 1st argument to be a positive integer, got {n}"
-                        ),
-                            pos_op,
-                        ))
-                    } else {
-                        let mut shared_env = Environment::new();
-                        let f_as_var = f.body.closurize(&mut self.cache, &mut env, f.env);
-
-                        // Array elements are closurized to preserve laziness of data structures. It
-                        // maintains the invariant that any data structure only contain indices (that is,
-                        // currently, variables).
-                        let ts = (0..n_int)
-                            .map(|n| {
-                                mk_app!(f_as_var.clone(), Term::Num(n as f64)).closurize(
-                                    &mut self.cache,
-                                    &mut shared_env,
-                                    env.clone(),
-                                )
-                            })
-                            .collect();
-
-                        Ok(Closure {
-                            body: RichTerm::new(
-                                Term::Array(ts, ArrayAttrs::new().closurized()),
-                                pos_op_inh,
-                            ),
-                            env: shared_env,
-                        })
-                    }
-                } else {
-                    Err(EvalError::TypeError(
+                let Term::Num(ref n) = *t else {
+                    return Err(EvalError::TypeError(
                         String::from("Num"),
                         String::from("generate, 1st argument"),
                         arg_pos,
                         RichTerm { term: t, pos },
                     ))
+                };
+
+                if n < &Number::ZERO {
+                    return Err(EvalError::Other(
+                        format!(
+                            "generate: expected the 1st argument to be a positive number, got {n}"
```

Comment:
For the trace primop, I used the `builtin.trace` as the output prefix. Maybe this here should use `array.generate` as well? We're generally assuming that the functions in the stdlib are addressed by their full path (without trickery like `let`-rebinding).